File: NeoRemote/app/src/main/java/neoplayer/neoremote/Message.java
package neoplayer.neoremote;

import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class Message {
    private ByteBuffer byteBuffer;

    public Message() {
        byteBuffer = ByteBuffer.allocateDirect(1024).order(ByteOrder.LITTLE_ENDIAN);
        add(0);
    }

    public Message add(byte[] value) {
        byteBuffer.put(value);
        return this;
    }

    public Message add(boolean value) {
        byteBuffer.put((byte) (value ? 1 : 0));
        return this;
    }

    public Message add(int value) {
        byteBuffer.putInt(value);
        return this;
    }

    public Message add(String value) {
        byte[] bytes;
        try {
            bytes = value.getBytes("UTF-8");
        } catch (Exception ex) {
            bytes = new byte[0];
        }
        add(bytes.length);
        add(bytes);
        return this;
    }

    public Message add(VideoFile videoFile) {
        add(videoFile.videoFileID);
        add(videoFile.title);
        return this;
    }

    public Message add(EditTags editTags) {
        add(editTags.videoFileIDs.size());
        for (int videoFileID : editTags.videoFileIDs)
            add(videoFileID);
        add(editTags.tags.size());
        for (Map.Entry<String, String> tag : editTags.tags.entrySet()) {
            add(tag.getKey());
            add(tag.getValue());
        }
        return this;
    }

    public byte[] toArray() {
        // Write size
        int size = byteBuffer.position();
        byteBuffer.position(0);
        byteBuffer.putInt(size);
        byteBuffer.position(size);

        byteBuffer.flip();
        byte[] arr = new byte[byteBuffer.remaining()];
        byteBuffer.get(arr);
        return arr;
    }

    public Message(InputStream stream) throws IOException {
        byte[] buffer = new byte[4];
        int size = 4;
        int used = 0;
        boolean first = true;
        while (used < size) {
            int block = stream.read(buffer, used, buffer.length - used);
            if (block == -1)
                throw new IOException();
            used += block;
            if ((first) && (used == size)) {
                first = false;
                size = ByteBuffer.wrap(buffer).order(ByteOrder.LITTLE_ENDIAN).getInt() - 4;
                buffer = new byte[size];
                used = 0;
            }
        }
        byteBuffer = byteBuffer.wrap(buffer).order(ByteOrder.LITTLE_ENDIAN);
    }

    public boolean getBool() {
        return byteBuffer.get() != 0;
    }

    public int getInt() {
        return byteBuffer.getInt();
    }

    public ArrayList<Integer> getInts() {
        int count = getInt();
        ArrayList<Integer> ints = new ArrayList<>();
        for (int ctr = 0; ctr < count; ++ctr) {
            ints.add(getInt());
        }
        return ints;
    }

    public String getString() {
        int size = getInt();
        if (size == -1)
            return null;

        byte[] bytes = new byte[size];
        byteBuffer.get(bytes, 0, size);
        try {
            return new String(bytes, "UTF-8");
        } catch (Exception ex) {
            return null;
        }
    }

    public VideoFile getVideoFile() {
        int videoFileID = getInt();
        String title = getString();
        int count = getInt();
        HashMap<String, String> tagValues = new HashMap<>();
        for (int ctr = 0; ctr < count; ++ctr)
            tagValues.put(getString(), getString());
        return new VideoFile(videoFileID, title, tagValues);
    }

    public ArrayList<VideoFile> getVideoFiles() {
        int count = getInt();
        ArrayList<VideoFile> videoFiles = new ArrayList<>();
        for (int ctr = 0; ctr < count; ++ctr) {
            videoFiles.add(getVideoFile());
        }
        return videoFiles;
    }

    public DownloadData getDownloadData() {
        String title = getString();
        int progress = getInt();
        return new DownloadData(title, progress);
    }

    public ArrayList<DownloadData> getDownloadDatas() {
        int count = getInt();
        ArrayList<DownloadData> downloadDatas = new ArrayList<>();
        for (int ctr = 0; ctr < count; ++ctr) {
            downloadDatas.add(getDownloadData());
        }
        return downloadDatas;
    }
}

File: NeoPlayer/Networking/Message.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using NeoPlayer.Models;

namespace NeoPlayer.Networking
{
	public class Message
	{
		readonly MemoryStream ms = new MemoryStream();

		public async static Task<Message> GetAsync(NeoSocket neoSocket)
		{
			var message = new Message();
			var ms = message.ms;

			ms.SetLength(0);
			var first = true;
			var size = 4;
			var buffer = new byte[1024];
			while (ms.Length < size)
			{
				var block = await neoSocket.ReadAsync(buffer, 0, (int)Math.Min(buffer.Length, size - ms.Length));
				ms.Write(buffer, 0, block);

				if ((first) && (ms.Length == size))
				{
					first = false;
					size = BitConverter.ToInt32(ms.GetBuffer(), 0);
				}
			}
			ms.Position = 4;

			return message;
		}

		public Message()
		{
			Add(0);
		}

		public void Add(byte[] value) => ms.Write(value, 0, value.Length);

		public void Add(bool playing) => Add(new byte[] { (byte)(playing ? 1 : 0) });

		public void Add(int value) => Add(BitConverter.GetBytes(value));

		public void Add(List<int> values)
		{
			Add(values.Count);
			values.ForEach(value => Add(value));
		}

		public void Add(long value) => Add(BitConverter.GetBytes(value));

		public void Add(string value)
		{
			if (value == null)
			{
				Add(-1);
				return;
			}
			var bytes = Encoding.UTF8.GetBytes(value);
			Add(bytes.Length);
			Add(bytes);
		}

		public void Add(VideoFile value)
		{
			Add(value.VideoFileID);
			Add(value.Title);
			Add(value.Tags.Count);
			foreach (var pair in value.Tags)
			{
				Add(pair.Key);
				Add(pair.Value);
			}
		}

		public void Add(List<VideoFile> values)
		{
			Add(values.Count);
			values.ForEach(value => Add(value));
		}

		public void Add(DownloadData value)
		{
			Add(value.Title);
			Add(value.Progress);
		}

		public void Add(List<DownloadData> values)
		{
			Add(values.Count);
			values.ForEach(value => Add(value));
		}

		public byte[] ToArray()
		{
			var result = ms.ToArray();
			Array.Copy(BitConverter.GetBytes(ms.Length), result, 4);
			return result;
		}

		public byte[] GetBytes(int count)
		{
			var result = new byte[count];
			ms.Read(result, 0, count);
			return result;
		}

		internal void Add(object value)
		{
			if (value is byte[])
				Add(value as byte[]);
			else if (value is bool)
				Add((bool)value);
			else if (value is int)
				Add((int)value);
			else if (value is long)
				Add((long)value);
			else if (value is string)
				Add(value as string);
			else if (value is VideoFile)
				Add(value as VideoFile);
			else if (value is List<VideoFile>)
				Add(value as List<VideoFile>);
			else if (value is DownloadData)
				Add(value as DownloadData);
			else if (value is List<DownloadData>)
				Add(value as List<DownloadData>);
			else if (value is List<int>)
				Add(value as List<int>);
			else
				throw new Exception($"Unknown type: {value?.GetType().Name ?? "<NULL>"}");
		}

		public bool GetBool() => GetBytes(1)[0] != 0;

		public int GetInt() => BitConverter.ToInt32(GetBytes(4), 0);

		public string GetString() => Encoding.UTF8.GetString(GetBytes(GetInt()));

		public EditTags GetEditTags()
		{
			var editTags = new EditTags { VideoFileIDs = new List<int>(), Tags = new Dictionary<string, string>() };

			var count = GetInt();
			for (var ctr = 0; ctr < count; ++ctr)
				editTags.VideoFileIDs.Add(GetInt());

			count = GetInt();
			for (var ctr = 0; ctr < count; ++ctr)
				editTags.Tags[GetString()] = GetString();

			return editTags;
		}
	}
}
